pkg/cache/json.go

```diff
@@ -58,31 +60,66 @@ func (q *JSON) ListBundles(ctx context.Context) ([]*api.Bundle, error) {
 }
 
 func (q *JSON) SendBundles(_ context.Context, s registry.BundleSender) error {
+	var keys []apiBundleKey
 	for _, pkg := range q.packageIndex {
-		channels := sets.KeySet(pkg.Channels)
-		for _, chName := range sets.List(channels) {
-			ch := pkg.Channels[chName]
-
-			bundles := sets.KeySet(ch.Bundles)
-			for _, bName := range sets.List(bundles) {
-				b := ch.Bundles[bName]
-				apiBundle, err := q.loadAPIBundle(apiBundleKey{pkg.Name, ch.Name, b.Name})
-				if err != nil {
-					return fmt.Errorf("convert bundle %q: %v", b.Name, err)
-				}
-				if apiBundle.BundlePath != "" {
-					// The SQLite-based server
-					// configures its querier to
-					// omit these fields when
-					// bundle path is set.
-					apiBundle.CsvJson = ""
-					apiBundle.Object = nil
-				}
-				if err := s.Send(apiBundle); err != nil {
-					return err
-				}
+		for _, ch := range pkg.Channels {
+			for _, b := range ch.Bundles {
+				keys = append(keys, apiBundleKey{pkg.Name, ch.Name, b.Name})
+			}
+		}
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		if keys[i].chName != keys[j].chName {
+			return keys[i].chName < keys[j].chName
+		}
+		if keys[i].pkgName != keys[j].pkgName {
+			return keys[i].pkgName < keys[j].pkgName
+		}
+		return keys[i].name < keys[j].name
+	})
+	var files []*os.File
+	var readers []io.Reader
+	for _, key := range keys {
+		filename, ok := q.apiBundles[key]
+		if !ok {
+			return fmt.Errorf("package %q, channel %q, key %q not found", key.pkgName, key.chName, key.name)
+		}
+		file, err := os.Open(filename)
+		if err != nil {
+			return fmt.Errorf("failed to open file for package %q, channel %q, key %q: %w", key.pkgName, key.chName, key.name, err)
+		}
+		files = append(files, file)
+		readers = append(readers, file)
+	}
+	defer func() {
+		for _, file := range files {
+			if err := file.Close(); err != nil {
+				logrus.WithError(err).WithField("file", file.Name()).Warn("could not close file")
 			}
 		}
+	}()
+	multiReader := io.MultiReader(readers...)
+	decoder := json.NewDecoder(multiReader)
+	index := 0
+	for {
+		var bundle api.Bundle
```

Comment:
Is it possible/would it help/is it safe (given we pass a pointer to `Send`) to declare `bundle` this outside the loop and reuse it?

Reply:
If `api.Bundle` has an optional field and on `index=0` we set it, but `index=1` does not ... what does the content of the variable look like?

Reply:
If it's logically correct, this is a good idea - we see an even larger reduction with it:

```
$ go tool pprof -diff_base /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.018.pb.gz
File: opm
Build ID: 6f0d15b7c8b7bcd22779237e2d78142137ab2ef4
Type: inuse_space
Time: Sep 6, 2023 at 9:26am (MDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for -662.18kB, 9.23% of 7175.57kB total
Dropped 2 nodes (cum <= 35.88kB)
Showing top 10 nodes out of 79
      flat  flat%   sum%        cum   cum%
-1114.54kB 15.53% 15.53% -1114.54kB 15.53%  encoding/json.(*decodeState).literalStore
 -557.26kB  7.77% 23.30%  -557.26kB  7.77%  github.com/russross/blackfriday/v2.init
 -536.37kB  7.48% 30.77%  -536.37kB  7.48%  k8s.io/apimachinery/pkg/api/meta.init
  518.02kB  7.22% 23.55%   518.02kB  7.22%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
     515kB  7.18% 16.38%      515kB  7.18%  regexp.makeOnePass.func1
     513kB  7.15%  9.23%      513kB  7.15%  golang.org/x/net/http2/hpack.newInternalNode
  512.38kB  7.14%  2.09%   512.38kB  7.14%  github.com/operator-framework/operator-registry/cmd/opm/index.newIndexDeprecateTruncateCmd
 -512.28kB  7.14%  9.23% -1626.82kB 22.67%  github.com/operator-framework/operator-registry/pkg/cache.(*JSON).Load
 -512.16kB  7.14% 16.36%  -512.16kB  7.14%  net.parseNSSConf
  512.02kB  7.14%  9.23%   512.02kB  7.14%  k8s.io/api/batch/v1.init
```

Reply:
Seems to lead to a 500KiB savings but the source is unclear ... 

```
$ go tool pprof -diff_base /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz /home/stevekuznetsov/pprof/pprof.opm.alloc_objects.alloc_space.inuse_objects.inuse_space.019.pb.gz
File: opm
Build ID: 6f0d15b7c8b7bcd22779237e2d78142137ab2ef4
Type: inuse_space
Time: Sep 6, 2023 at 9:26am (MDT)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for -662.16kB, 9.23% of 7175.57kB total
Dropped 5 nodes (cum <= 35.88kB)
Showing top 10 nodes out of 81
      flat  flat%   sum%        cum   cum%
-1114.54kB 15.53% 15.53%  -602.49kB  8.40%  encoding/json.(*decodeState).literalStore
 -557.26kB  7.77% 23.30%  -557.26kB  7.77%  github.com/russross/blackfriday/v2.init
 -536.37kB  7.48% 30.77%  -536.37kB  7.48%  k8s.io/apimachinery/pkg/api/meta.init
  518.02kB  7.22% 23.55%   518.02kB  7.22%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
     515kB  7.18% 16.38%      515kB  7.18%  regexp.makeOnePass.func1
     513kB  7.15%  9.23%      513kB  7.15%  golang.org/x/net/http2/hpack.newInternalNode
  512.38kB  7.14%  2.09%   512.38kB  7.14%  github.com/operator-framework/operator-registry/cmd/opm/index.newIndexDeprecateTruncateCmd
 -512.28kB  7.14%  9.23% -1626.82kB 22.67%  github.com/operator-framework/operator-registry/pkg/cache.(*JSON).Load
 -512.16kB  7.14% 16.36%  -512.16kB  7.14%  net.parseNSSConf
  512.05kB  7.14%  9.23%   512.05kB  7.14%  encoding/json.unquoteBytes
```

Reply:
> If api.Bundle has an optional field and on index=0 we set it, but index=1 does not ... what does the content of the variable look like?

Right, we'd need to `.Reset()` it. But it looks like that just [re-initializes](https://github.com/operator-framework/operator-registry/blob/d02b0b68acab6bb218b431bed70b3703ffc168ab/pkg/api/registry.pb.go#L392) to an empty `Bundle`, so maybe unhelpful.